Simplify backup code and add a pagecache parameter to neo4j-admin backup

chrisvest opened the pull request:
* Collapsed inheritance and implementation hierarchies.
* Collapsed factory-provider-builder chains.
* Remove unused code.
* Shorten names of classes.
* Hide classes that were accidentally added to the public API by making them package private.
* Avoid using service-loading for getting default instances of things. Only alternatives are service-loaded.
* Collapse the family of helper classes around configuring the online backup admin command into a single class.

changelog: [3.4] Add `--pagecache` parameter to the `neo4j-admin backup` command, with a default value of 8 MiB that now overrides the page cache setting in `neo4j.conf`.

phughk committed:
Backups on blockdevices

Adds support/fixes tests for performing backups on blockdevices

phughk committed:
Backups on block devices - addressed PR comments

Addressed comments by Chris Vest and Andrew Kerr

chrisvest committed:
Simplify backup code and add a pagecache parameter to neo4j-admin backup.

* Collapsed inheritance and implementation hierarchies.
* Collapsed factory-provider-builder chains.
* Remove unused code.
* Shorten names of classes.
* Hide classes that were accidentally added to the public API by making them package private.
* Avoid using service-loading for getting default instances of things. Only alternatives are service-loaded.
* Collapse the family of helper classes around configuring the online backup admin command into a single class.

chrisvest commented:
Currently marked `NOT READY FOR MERGE` because there are still more things that can be done.

chrisvest committed:
Merge branch '3.4-newCAPI' into 3.4-force-small-backup-pagecache

phughk commented on enterprise/backup/src/main/java/org/neo4j/backup/BackupFlowFactory.java:
```diff
@@ -19,51 +19,60 @@
-Backup flows are iterate through backup strategies and make sure at least one of them is a valid backup. Handles cases when that isn't possible.
-This factory helps in the construction of them
+/**
+ * Backup flows are iterate through backup strategies and make sure at least one of them is a valid backup. Handles
```
There was a grammar error in the original, `are` can be removed.

chrisvest committed:
Add more debug utilities that turned out to be useful.

chrisvest committed:
Make the ClusterRule assign a dedicated thread group for each member in the test clusters.

This makes it easier to debug problems in the cluster tests, beucase the debugger can now show what cluster member each thread belongs to.

chrisvest committed:
Add a helpful comment to RemoteStore about why we always ignore the transaction log directory setting during store copy.

chrisvest committed:
Fix some checkstyle problems.

chrisvest committed:
Merge branch '3.4' into 3.4-force-small-backup-pagecache

chrisvest committed:
Hide classes that really should not be public from the public API.

chrisvest committed:
Remove the backup implementation from public view by putting it in an impl package.

chrisvest committed:
Cleanups around backup code.

chrisvest committed:
Fix a bug in FileMoveProvider

Moving a single file, as opposed to a directory of files, was not
handled correctly.

chrisvest committed:
Fix weird clustering test.

chrisvest committed:
Remove almost-mock-only that is flaky due to combining mocks and threads.

chrisvest committed:
Make AdminTool read the NEO4J_ environment variables through mutable static fields, so tests could potentially modify them in-process.

chrisvest committed:
Check specifically for the `neostore` file for existing backups.

This allows us to download a backup into a directory that already has
contents, as long as those contents don't appear to be a Neo4j store.

This is important for supporting backup in clusters that use encrypted
communications based on pre-shared key files.

chrisvest committed:
Fix a bug in the record format selector.

If the page cache was using a different file system than the rest of the
database, the record format selector would not be able to find the
`neostore` file when inspecting the store format version, because it was
looking through the wrong file system.

chrisvest commented:
@MishaDemianenko @phughk Please take a look at this, _and_ the referenced PRs in the other repositories.

MishaDemianenko commented on enterprise/backup/src/main/java/org/neo4j/backup/impl/BackupCopyService.java:
```diff
@@ -0,0 +1,141 @@
+        }
+    }
+
+    public void clearLogs( Path neo4jHome )
```
why would we need to remove logs(i saw that according to call path's that when we had successful backups, but why)?

chrisvest replied:
I don't know. Can you answer, @phughk ?

phughk replied:
Ah, that is an actual error on my part, sorry! It should be `clearIdFiles` and have the following implementation
https://github.com/neo4j/neo4j/blob/3.2/enterprise/backup/src/main/java/org/neo4j/backup/BackupService.java#L437

chrisvest replied:
I'll have a fix for this ready soon.

MishaDemianenko commented on enterprise/backup/src/main/java/org/neo4j/backup/impl/BackupSupportingClassesFactoryProvider.java:
```diff
@@ -17,53 +17,50 @@
+                load( BackupSupportingClassesFactoryProvider.class ).spliterator(), false );
+        // Inject the default provider into the stream, so it participates in sorting by priority.
+        providers = Stream.concat( providers, Stream.of( new BackupSupportingClassesFactoryProvider() ) );
+        return providers.sorted( ( l, r ) -> r.getPriority() - l.getPriority() );
```
Comparator.comparingInt( BackupSupportingClassesFactoryProvider::getPriority )

MishaDemianenko commented on enterprise/backup/src/main/java/org/neo4j/backup/impl/OnlineBackupCommandBuilder.java:
```diff
@@ -0,0 +1,358 @@
+        }
+        catch ( IOException e )
+        {
+            e.printStackTrace();
```
Does not look like a good handling of exception? should we be able to continue if this fails?

phughk replied:
It's part of a builder, so it shouldn't be throwing a checked exception but having it wrapped as unchecked should be fine. Simply printing out the trace is wrong, I admit I overlooked that.

MishaDemianenko commented on enterprise/backup/src/main/java/org/neo4j/backup/impl/OnlineBackupCommandBuilder.java:
```diff
@@ -0,0 +1,358 @@
+    private static String[] args( String... args )
+    {
+        List<String> cleanedArgs = Stream.of( args )
+                .filter( Objects::nonNull )
```
this whole thing can be replaced with 
`Stream.of( args )
                .filter( StringUtils::isNotEmpty )
                .toArray( String[]::new )` right?

phughk replied:
Yes, thanks!

MishaDemianenko commented on enterprise/backup/src/main/java/org/neo4j/backup/impl/OnlineBackupCommandBuilder.java:
```diff
@@ -0,0 +1,358 @@
+        {
+            return "";
+        }
+        File configFile = neo4jHomeFromTarget( backupTarget ).resolve( "../additional_neo4j.conf" ).toFile();
```
what is additional_neo4j.conf and why do we need it? how its correlate with other tools behavior?

chrisvest replied:
Can you answer this one, @phughk ?

phughk replied:
The idea behind this code is that you could supply an `Config additionalConfig` that would override defaults. This method in the builder would physically write that config to a file (that is hardcoded as additional_neo4j.conf) and add it as a parameter.
I don't think it is used anywhere at the moment, so if you think it should be thrown out then please go ahead.

MishaDemianenko replied:
while the idea to have an ability to provide additional config can be a good one or not, its important that all tools that we have should have same capabilities in terms how they can be configured, where they store their logs, etc and having something that specific for a backup can be not that good idea, but i will leave it for you guys

chrisvest replied:
This feature is used in the related blockdevice PR, so let's keep it.

MishaDemianenko commented on enterprise/backup/src/test/java/org/neo4j/backup/impl/BackupProtocolServiceIT.java:
```diff
@@ -789,12 +794,13 @@ public void shouldHandleBackupWhenLogFilesHaveBeenDeleted() throws Exception
             throws IOException
     {
-        final FileFilter logFileFilter = pathname -> pathname.getName().contains( "logical" );
+        final Predicate<Path> logFileFilter = p -> p.getFileName().toString().contains( "logical" );
```
would we delete any "logical" file? shouldn't it use our log files abstraction instead?

MishaDemianenko commented on enterprise/com/src/test/java/org/neo4j/com/storecopy/FileMoveProviderTest.java:
```diff
@@ -0,0 +1,246 @@
+    }
+
+    @Test
+    public void filesAreMovedBeforeDirectories() throws IOException // TODO doesnt test anything maybe
```
TODO

MishaDemianenko reviewed (changes requested):
Only couple of small comments here and there, looks good otherwise

chrisvest committed:
Merge branch '3.4' into 3.4-force-small-backup-pagecache

chrisvest committed:
Verified that test tests something; removed TODO.

chrisvest committed:
Use NeoStoreDataSource to find log files in test.

chrisvest committed:
OnlineBackupCommandBuilder cleanup

chrisvest committed:
Small code cleanup

chrisvest committed:
Revert "Small code cleanup"

This reverts commit dae24fba584c5adf8596211b921ad7bc620fc766.

chrisvest committed:
Small cleanup and clarification of the provider sorting in BackupSupportingClassesFactoryProvider.

chrisvest committed:
Fix OnlineBackupContextBuilderTest after recent product changes.

chrisvest committed:
Backup should clear the store id files.

Taking an online backup effectively looks like a system crash, followed by a recovery.
In such cases, the id files are not trustworthy anyway.

More importantly, however, a backup may be used to seed a cluster, and in such cases the id files must not contain any free ids that can be reused.
The reason is that id-reuse is an uncoordinated operation in clustering, and therefor no two cluster members can be allowed to have overlapping sets of free ids.
Overlapping free id sets in a cluster can, with unfortunate timings, potentially lead to inconsistencies.

chrisvest committed:
Merge branch '3.4' into 3.4-force-small-backup-pagecache

chrisvest committed:
Merge branch '3.4' into 3.4-force-small-backup-pagecache

chrisvest committed:
Update new license headers.
